llmfoundry/command_utils/data_prep/convert_delta_to_mds.py

```diff
@@ -0,0 +1,167 @@
+# Copyright 2024 MosaicML LLM Foundry authors
+# SPDX-License-Identifier: Apache-2.0
+
+import json
+import logging
+import os
+import tempfile
+from typing import Callable, Optional
+
+import numpy as np
+from streaming import MDSWriter
+
+from llmfoundry.command_utils.data_prep.convert_delta_to_json import (
+    _check_imports,
+    fetch_DT,
+    format_tablename,
+    get_columns_info,
+    validate_and_get_cluster_info,
+)
+
+logger = logging.getLogger(__name__)
+
+
+def get_conversion_config(
+    columns: list[str],
+    provided_dtypes: Optional[dict],
+) -> tuple[dict, Callable]:
+    """If no dtypes is provided, attempts to infer config based on column names.
+
+    Args:
+        columns (List[str]): The list of column names.
+        provided_dtypes (Optional[Dict]): The provided dtypes.
+    """
+    if provided_dtypes is not None:
+        convert_x = lambda x: {
+            k: np.array(v, dtype=provided_dtypes.get(k)) for k, v in x.items()
+        }
+        return provided_dtypes, convert_x
+
+    if len(columns) != 1:
+        raise ValueError(
+            'Unable to infer dtypes from columns and no dtypes provided.',
+        )
+
+    if 'turns' in columns[0]:
+        logging.info('Identified IFT data')
+        dtypes = {
+            'input_ids': 'ndarray',
+            'attention_mask': 'ndarray',
+            'labels': 'ndarray',
+        }
+        convert_x = lambda x: {
+            # join the turns into a single array
+            'input_ids':
+                np.concatenate([
```

Comment:
Q: Is this correct / how to convert multi-turn chat to MDS?

Reply:
You should try to factor and reuse code from here if possible: https://github.com/mosaicml/llm-foundry/blob/41e02d778c22185dbcd3fd4fedd64ff264883ee9/llmfoundry/command_utils/data_prep/convert_finetuning_dataset.py#L173-L267

Reply:
The tokenized IFT/CHAT output is the output of the HF tokenizer (input_ids, attention_mask, labels). Just checked, multi-turn chat gets concatenated / formatted into a single turn. Assuming single-turn input for this function and removing multi-turn concat

Reply:
Took a look at the code you linked. IMO, not easy to re-factor + re-use because most of the logic for formatting examples, which is assumed to happen before data is written to Delta table. Like mentioned in the previous comment, it looks like multi-turn CHAT data gets concatenated into a single turns array (similar to the linked logic). The last remaining step is writing as MDS format, which is what this PR adds.